src/keccak.js

```diff
@@ -57,54 +61,110 @@
   /**
    * Rotate left Uint32Array(2)
    *
-   * FIXME Allocate a new array. Do not modify the array provided as an input.
+   * TODO Create unit tests for this function.
    *
-   * @param a64 {Uint32Array} of size 2
-   * @param n
-   * @returns {Uint32Array} of size 2
+   * @param a64 {Uint32Array} of size 2 representing a Uint64
+   * @param n {Number} number of bits to rotate left
+   * @returns {Uint32Array} of size 2 representing a Uint64
    */
   function rotl64(a64, n) {
     n = (n % 64) | 0;
-
-    var lo32 = a64[LO32];
-    var hi32 = a64[HI32];
-
+    var r64 = new Uint32Array(2);
     if (n === 0) {
-      return a64;
+      r64[LO32] = a64[LO32];
+      r64[HI32] = a64[HI32];
     } else if (n === 32) {
-      a64[HI32] = lo32;
-      a64[LO32] = hi32;
+      r64[LO32] = a64[HI32];
+      r64[HI32] = a64[LO32];
     } else if (n < 32) {
-      a64[HI32] = (lo32 << n) | (hi32 >> (32 - n));
-      a64[LO32] = (hi32 << n) | (lo32 >> (32 - n));
+      r64[LO32] = (a64[LO32] << n) | (a64[HI32] >> (32 - n));
+      r64[HI32] = (a64[HI32] << n) | (a64[LO32] >> (32 - n));
     } else if (n > 32) {
-      n = (32 - n) | 0;
-      a64[LO32] = (lo32 << n) | (hi32 >> (32 - n));
-      a64[HI32] = (hi32 << n) | (lo32 >> (32 - n));
+      n = (n - 32) | 0;
+      r64[LO32] = (a64[HI32] << n) | (a64[LO32] >> (32 - n));
+      r64[HI32] = (a64[LO32] << n) | (a64[HI32] >> (32 - n));
```

Comment:
I need to read more about this since I'm sure I'm wrong about everything, but I believe the fact that we're accessing a number in a uint32 array doesn't override the fact that bitwise ops in javascript occur on 32-bit signed numbers. 

So if our a64[LO32] happens to have a 1 in the sign bit position, and we do a sign preserving right shift, we're going to get the wrong result?

Recommend always using zero-fill right shift (>>>) in all of our operations, since I believe we're basically only interested in "unsigned 64 bit integers" (i think).

http://www.ecma-international.org/ecma-262/5.1/#sec-11.7.2